tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc

```diff
@@ -4489,16 +4489,21 @@ class ConvertTensorScatterOp : public OpRewritePattern<OpTy> {
         op.updates().getType().template dyn_cast<RankedTensorType>();
 
     if (!tensor_ty || !indices_ty || !updates_ty) return failure();
-    // Last dimension of the indices needs to known at compile time for
-    // computation of the 'update_window_dims' attribute in the dimensions
-    // struct.
-    int64_t num_index_dims = indices_ty.getShape().back();
-    if (ShapedType::isDynamic(num_index_dims)) return failure();
 
     int64_t tensor_rank = tensor_ty.getRank();
     int64_t indices_rank = indices_ty.getRank();
     int64_t updates_rank = updates_ty.getRank();
 
+    // Handling special case with rank0 and rank1 indices, making them rank2
+    std::vector<int64_t> indices_shape = indices_ty.getShape();
+    if(indices_rank == 0) indices_shape = {1,1};
+    else if(indices_rank == 1) indices_shape.push_back(1);
```

Comment:
What's the rationale for this logic?

From the shape inference function, it looks like the minor dim is used for vector as well. 
https://github.com/tensorflow/tensorflow/blob/81aa90bf6c86b6af7d9d37d08e92915ee6096801/tensorflow/core/framework/common_shape_fns.cc#L2366

Reply:
I ran through an error lowering down BERT training to MHLO from IREE, and turns out they got the windows_dim to be negative when computing. Talking to Rob from IREE he mentioned that "That the minimum dims the non-resource variant handle is 2 dim", so we need to handle it as such

Reply:
The MLIR op that broke the lowering, will add some generalization of this into the test code :) 
```    
%1437 = "tf.TensorScatterAdd"(%arg0, %arg1, %arg2) : (tensor<100x768xf32>, tensor<512xi32>, tensor<512x768xf32>) -> tensor<*xf32>
```

Reply:
so basically if we get indices.shape = [N], we'd make it [N,1]

Reply:
That seems like an invalid op. 
https://github.com/tensorflow/tensorflow/blob/7c96db88e54cc417d416aa1aa405145e15ca0731/tensorflow/compiler/tf2xla/lib/scatter.cc#L51

Reply:
hmm is it perhaps because, when they check it is [512,1], we do remove_suffix as below:
https://github.com/tensorflow/tensorflow/blob/7c96db88e54cc417d416aa1aa405145e15ca0731/tensorflow/compiler/tf2xla/lib/scatter.cc#L58
🤔

Reply:
Where do we add 1 in the suffix if the original indices were [512]?

I think, it will fail the following check and not generate the negative window_dim -510.

num_index_dims > buffer_shape.rank()

Reply:
I see I see, does that mean we should just add this check to the MLIR version?

Reply:
and does that mean there are some problem compiling down Transformer model for training?

Reply:
for reference this is the mlir line I got:
```
    %1437 = "tf.TensorScatterAdd"(%arg0, %arg1, %arg2) : (tensor<100x768xf32>, tensor<512xi32>, tensor<512x768xf32>) -> tensor<*xf32>
```

Reply:
and if we do add the check doesn't that mean we can't compile any 1d vector with indices and updates that has greater rank than the buffer?
For example:
<img width="495" alt="Screen Shot 2021-06-28 at 12 25 45 PM" src="https://user-images.githubusercontent.com/68087699/123692874-2d879700-d80c-11eb-9a5c-f7eec1b95659.png">
whose num_index_dim = 8, and buffer.rank = 1

Reply:
Maybe the XLA compiler take care of this issue in the higher level? that's why it doesn't have this problem in the XLA compiler?

Reply:
These indices need to be of rank 2 in that example. Please check the code that generated this op and that would probably require a fix.

TensorScatterAddOp doc mentions this check.
https://github.com/tensorflow/tensorflow/blob/4fb5c5f49d92edafe5883b154aa546a0c00c7d65/tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td#L17908

Reply:
Hey Smit, thanks for the guidance! 

I have a patch that handles it in the generation of tensorScatteradd which is part of decomposition of resourcescatteradd(the highest level I can see): https://github.com/raikonenfnu/tensorflow/commit/ac8750d297b1ffef7e94bf8b668bb46d0c2eebd5 

Any thoughts? :)

Reply:
Thanks for finding that. That seems in the right direction.

Why does that need to be applied only to tensors with rank less than two? I haven't checked the semantics but seems like it should be applied to all.

Reply:
Hmm good question, can we assume that once it has at least 2 dimension, it will have the proper indice.shape[-1] size? 🤔

Reply:
Hmm, I have been thinking about it, I think aside from rank-1 indices, other rank indices should be fine.
For example if we have a rank 2 indices as such:

indices = {[1,1],[2,4],[3,5]} : tensor<3x2xi32>, we would want num_of_index to be 2,  and 2 is already indices_shape.back()

Another rank-2 indices = {[1,2,3],[4,5,6],[7,8,9]} : tensor <3x3xi32>, in this case we would want num of index to be 3, and 3 is also already indices_shape.back()

Reply:
Sorry, didn't see this comment earlier.

Where did you see these examples? Looking at the XlaOpKernel implementation, it doesn't differentiate between ranks and always drops the last dim of indices in TensorScatterAdd.

Reply:
Hey @smit-hinsu, I am not too familiar with xlaopkernel implementations, but I think general TF doesn't take care of this case yet. For example if we have this it will work:
```python
    indices = tf.constant([[4], [3], [1], [7]]) // Becomes shape of [4,1]
    updates = tf.constant([9, 10, 11, 12])
    tensor = tf.ones([8], dtype=tf.int32)
    updated = tf.tensor_scatter_nd_add(tensor, indices, updates)
``` 

but in the case of 
```python
    indices = tf.constant([4, 3, 1, 7]) // Becomes shape of [4,]
    updates = tf.constant([9, 10, 11, 12])
    tensor = tf.ones([8], dtype=tf.int32)
    updated = tf.tensor_scatter_nd_add(tensor, indices, updates)
``` 
does not work and throw off 
```bash
tensorflow.python.framework.errors_impl.InvalidArgumentError
```

Reply:
That example made sense and it is correct that this is broken and we need to fix.

I was talking about the two examples you shared earlier. Where did you see this?

> indices = {[1,1],[2,4],[3,5]} : tensor<3x2xi32>, we would want num_of_index to be 2, and 2 is already indices_shape.back()
> 
> Another rank-2 indices = {[1,2,3],[4,5,6],[7,8,9]} : tensor <3x3xi32>, in this case we would want num of index to be 3, and 3 is also already indices_shape.back()

I looked at op definitions for TensorScatterAddOp and ResourceScatterAddOp in tf_generated_ops.td and we need to reshape indices for all cases even scalar and higher ranked indices. See 'indices.shape + ref.shape[1:]' for the resource op and 'indices.shape[:-1] + tensor.shape[indices.shape[-1]:]' for the other case. So indices.shape[-1] should always be 1. Please update the other PR you have.

Reply:
Ahh that makes sense! Thanks for the guidance @smit-hinsu! I'll update the other PR :)

Reply:
Hey @smit-hinsu, I pushed the TD-based change onto this PR. :)

Reply:
Extra Note: I have tested compiling Tensorflow BERT training graph all the way down to linalg generic with this patch, and it works :)